src/main/java/com/monumental/services/AwsS3Service.java

```diff
@@ -24,50 +29,77 @@
      */
-    public String storeObject(String bucketName, String objectKey, File file) {
-        String objectUrl = "";
-
-        AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
-                            .withRegion(bucketRegion)
-                            .build();
-
+    public String storeObject(String objectKey, File file) throws SdkClientException {
+        objectKey = generateUniqueKey(objectKey);
         try {
-            if (!s3Client.doesObjectExist(bucketName, objectKey)) {
-                s3Client.putObject(bucketName, objectKey, file);
-            }
-
-            objectUrl = getObjectUrl(bucketName, objectKey);
-        } catch (AmazonServiceException e) {
+            s3Client.putObject(bucketName, objectKey, file);
+            return getObjectUrl(objectKey);
+        } catch (SdkClientException e) {
             System.out.println("Error attempting to access S3 Bucket: " + bucketName + " and Object: " + objectKey);
-            System.out.println(e.getErrorMessage());
+            System.out.println(e.getMessage());
+            throw e;
         }
-
-        return objectUrl;
     }
 
     /**
      * Method to generate the Object URL for a specified bucketName and objectKey
-     * @param bucketName - The name of the Bucket for the Object URL
      * @param objectKey - The Object Key for the Object
      * @return String - The Object URL for the Object
      */
-    public static String getObjectUrl(String bucketName, String objectKey) {
-        return httpsProtocol + bucketName + s3Domain + objectKey;
+    public static String getObjectUrl(String objectKey) {
+        try {
+            return httpsProtocol + bucketName + s3Domain + URLEncoder.encode(
+                    objectKey,
+                    StandardCharsets.UTF_8.toString()
+            );
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    public static boolean isUniqueKey(String objectKey) {
+        try {
+            return !s3Client.doesObjectExist(bucketName, objectKey);
+        } catch (AmazonServiceException e) {
+            System.out.println("Error attempting to access S3 Bucket: " + bucketName + " and Object: " + objectKey);
+            System.out.println(e.getErrorMessage());
+            return true;
+        }
+    }
+
+    public static String generateUniqueKey(String objectKey) {
+        if (isUniqueKey(objectKey)) return objectKey;
+        else {
+            String checkRegex = "/\\([0-9]*\\)/$";
+            String captureRegex = "/\\(([0-9]*)\\)/$";
+            // If the key already ends with "(1)", for example
+            if (objectKey.contains(checkRegex)) {
+                Integer number = Integer.parseInt(Pattern.compile(captureRegex).matcher(objectKey).group(0)) + 1;
+                objectKey = objectKey.replaceAll(checkRegex, "(" + number + ")");
```

Comment:
What if someone uploads an image with the name `image (1).png` but we already have `image (2).png` in S3?

Reply:
This now is in a loop until it finds a unique key